linked_list/linked_list.py

```diff
@@ -1,103 +1,191 @@
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        current = self.head
+
+        if not current:
+            return None 
+
+        while current:
+            if current.next == None:
+                return current.value
+            current = current.next 
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) ?
     def add_last(self, value):
-        pass
+        new_last = Node(value)
+        if self.head == None:
+            self.head = new_last
+        else:
+            current = self.head
+
+            while current.next != None:
+                current = current.next
+            current.next = new_last
 
     # method to return the max value in the linked list
     # returns the data value and not the node
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_max(self):
-        pass
+        current = self.head 
+        max_ct = 0 # max char count var
+        
+        if not current: 
+            return None
+
+        while current != None:
+            if max_ct < current.value:
+                max_ct = current.value
+            current = current.next
+        return max_ct
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        current = self.head
+
+        if current != None:
+            if current.value == value: 
+                self.head = current.next 
+                current = None 
+                return
+
+        while current != None: 
+            if current.value == value: 
+                break 
+            prev = current
+            current = current.next
+
+        if current == None: 
+            return
+
+        prev.next = current.next 
+        current = None
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def visit(self):
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
+    def visit(self): # built by chris/ada
         helper_list = []
         current = self.head
 
-        while current:
+        while current: 
             helper_list.append(str(current.value))
             current = current.next
         
         print(", ".join(helper_list))
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def reverse(self):
-        pass
-  
-    ## Advanced/ Exercises
-    # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def find_middle_value(self):
-        pass
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) ?
+    def reverse(self): 

```

Comment:
👍 However the space complexity is O(1) because you're not making any new lists of nodes